Implemented input (#2779)
statelyai/xstate: docs/guides/input.md

> # Input <Badge text="4.25+" />

Reviewer: ```suggestion
# Input

_Since version 4.27.0_
```

[:rocket: Quick Reference](#quick-reference)

> When you're building software, it's common to split your code up into small pieces. These might be functions in a node process, or components in a React app. These small pieces often need to declare **dependencies** - bits of information/code that they're not responsible for, but that they rely on to serve their purpose.

Reviewer: ```suggestion
When you're building software, it's common to split your code up into small pieces. These might be functions in a node process, or framework components. These small pieces often need to declare **dependencies** - bits of information/code that they're not responsible for, but that they rely on to serve their purpose.
```

The best way to declare dependencies to an XState machine is called `input`. You can specify input like so:

```ts
const machine = createMachine(
  {
    entry: (context, event, meta) => {
      console.log(meta.state.input.lyric); // 'In west Philadelphia, born and raised'
    }
  },
  {
    input: {
      lyric: 'In west Philadelphia, born and raised'
    }
  }
);
```

Or, by using the `.withConfig` method:

```ts
const machineWithDifferentInput = machine.withConfig({
  input: {
    lyric: 'In the playground, where I spend most of my days'
  }
});
```

The input is then available in any action/service/guard, in the `meta` argument.

```ts
const machine = createMachine(
  {
    entry: [
      (context, event, meta) => {
        console.log(meta.state.input.foo);
      }
    ],
    invoke: {
      src: (context, event, meta) => {
        console.log(meta.state.input.foo);
      },
      onDone: {
        cond: (context, event, meta) => {
          console.log(meta.state.input.foo);
          return false;
        }
      }
    }
  },
  {
    input: {
      foo: 'bar'
    }
  }
);
```

> Unlike [context](./context.md), inputs cannot be assigned to from _inside_ the machine. If you do need to mutate the input locally, consider putting it in context instead.

Reviewer: ```suggestion
Unlike [context](./context.md), inputs cannot be assigned to from _inside_ the machine. If you do need to mutate the input locally, consider putting it in `context` instead.
```

The best way to declare dependencies to an XState machine is called `input`. You can specify input like so:

```ts
const machine = createMachine(
  {
    entry: (context, event, meta) => {
      console.log(meta.state.input.lyric); // 'In west Philadelphia, born and raised'
    }
  },
  {
    input: {
      lyric: 'In west Philadelphia, born and raised'
    }
  }
);
```

Or, by using the `.withConfig` method:

```ts
const machineWithDifferentInput = machine.withConfig({
  input: {
    lyric: 'In the playground, where I spend most of my days'
  }
});
```

The input is then available in any action/service/guard, in the `meta` argument.

```ts
const machine = createMachine(
  {
    entry: [
      (context, event, meta) => {
        console.log(meta.state.input.foo);
      }
    ],
    invoke: {
      src: (context, event, meta) => {
        console.log(meta.state.input.foo);

>       },

Reviewer 2: Since `meta.state` is currently not available to invoke creators this should be removed from an example and from other places mentioning this.

Author: Isn't it? Ah, crap

Reviewer 2: That is not hard to fix, but I... don't want to 😅 this is pending on the outcome of this: https://github.com/statelyai/xstate/pull/2779#discussion_r749321830 . I think we should decide on this asap - as it would be bad to introduce a new thing into a potentially deprecated "argument". If we can spare people from refactoring this in the future this would be awesome.

      onDone: {
        cond: (context, event, meta) => {
          console.log(meta.state.input.foo);
          return false;
        }
      }
    }
  },
  {
    input: {
      foo: 'bar'
    }
  }
);
```

Unlike [context](./context.md), inputs cannot be assigned to from _inside_ the machine. If you do need to mutate the input locally, consider putting it in `context` instead.

### Inputs that change over time

You can change an input on a service that's running by calling `service.input()`.

```ts
const service = interpret(machine).start();

service.input({
  foo: 'bar'
});
```

Under the hood, this sends a `xstate.input` event which XState recognises internally and uses to update the input.

### Listening for changes

You can listen for changes by listening for the `xstate.input` event:

```ts
const machine = createMachine({
  on: {
    'xstate.input': {
      actions: (context, event, meta) => {
        console.log(meta.state.input);
      }
    }
  }
});
```

Note that `meta` will be updated _before_ the actions and guards that listen for this transition are executed. This means you don't need to read the updated input from the event:

> ```ts {5-6,8-9}

Reviewer 2: what do those curly braces mean here? it's the first time I see this and I can't find any reference for this in https://github.github.com/gfm/#fenced-code-blocks

Author: They're for highlighting lines of code, used elsewhere in the repo

Reviewer: Yep, they're for Vuepress

# Input <Badge text="4.25+" />

[:rocket: Quick Reference](#quick-reference)

> When you're building software, it's common to split things up into small pieces. These might be functions in a node process, or components in a React app. These small pieces often need to declare **dependencies** - bits of information/code that they're not responsible for, but that they rely on to serve their purpose.

Reviewer 3: > things

What do you mean by “things” here? Is it “code”? “the project structure”? Something else? You give some examples in the next line, could we summarise the “things” with something more specific?

Author: ```suggestion
When you're building software, it's common to split your code up into small pieces. These might be functions in a node process, or components in a React app. These small pieces often need to declare **dependencies** - bits of information/code that they're not responsible for, but that they rely on to serve their purpose.
```

The best way to declare dependencies to an XState machine is called `input`. You can specify input like so:

```ts
const machine = createMachine(
  {

>     entry: (context, event, meta) => {
>       console.log(meta.state.input.lyric); // 'In west philadelphia, born and raised'
>     }
>   },
>   {
>     input: {
>       lyric: 'In west philadelphia, born and raised'

Reviewer 2: ```suggestion
    entry: (context, event, meta) => {
      console.log(meta.state.input.lyric); // 'In west Philadelphia, born and raised'
    }
  },
  {
    input: {
      lyric: 'In west Philadelphia, born and raised'
```

    }
  }
);
```

Or, by using the `.withConfig` method:

```ts
const machineWithDifferentInput = machine.withConfig({
  input: {

>     lyric: 'In the playground, where I spend most of my days'

Reviewer 3: Should this use the same lyric example as above, as it’s two different ways to achieve the same goal?

Author: No, because the inference is you can override it.

  }
});
```

> The input is then available in any action/service/guard, in the `meta` property.

Reviewer 2: ```suggestion
The input is then available in any action/service/guard, in the `meta` argument.
```

      }
    ],
    invoke: {
      src: (context, event, meta) => {
        console.log(meta.state.input.foo);
      },
      onDone: {
        cond: (context, event, meta) => {
          console.log(meta.state.input.foo);
          return false;
        }
      }
    }
  },
  {
    input: {
      foo: 'bar'
    }
  }
);
```

Unlike [context](./context.md), inputs cannot be assigned to from _inside_ the machine. If you do need to mutate the input locally, consider putting it in context instead.

### Inputs that change over time

You can change an input on a service that's running by calling `service.input()`.

```ts
const service = interpret(machine).start();

/**
 * It doesn't have to be the whole input -
 * it can be a partial
 */
service.input({
  foo: 'bar'
});
```

Under the hood, this sends a `xstate.input` event which XState recognises internally and uses to update the input.

### Listening for changes

You can listen for changes by listening for the `xstate.input` event:

```ts
const machine = createMachine({
  on: {
    'xstate.input': {
      actions: (context, event, meta) => {
        console.log(meta.state.input);
      }
    }
  }
});
```

> Note that `meta` will be updated _before_ actions and guards that listen for this transition fire. This means you don't need to read the updated input from the event:

Author: 😂 

```suggestion
Note that `meta` will be updated _before_ the actions and guards that listen for this transition are executed. This means you don't need to read the updated input from the event:
```

```ts {5-6,8-9}
const machine = createMachine({
  on: {
    'xstate.input': {
      actions: (context, event, meta) => {

>         // BAD

Author: It's bad because then you need to type the event properly. I'm comfortable here with recommending an approach without explaining why.

>         // BAD

Reviewer: ```suggestion
        // ⚠️ Avoid - prefer `meta.state.input` instead
```

        console.log(event.input);

>         // GOOD

Reviewer: ```suggestion
        // ✅
```

        console.log(meta.state.input);
      }
    }
  }
});
```

## Usage

### Frontend

> 'Dependencies which change over time', otherwise known as **props**, is a common pattern in modern frontend frameworks. Input allows you to use props in your state machines. See the docs on each individual framework to learn more:

Author: It's common in frontend, but not in backend. I figured I'd give a short definition here to clarify it.

- [React](../recipes/react.md#syncing-data-with-input)
- Vue (Not yet complete)
- Svelte (Not yet complete)

## Examples

### Listening for data from a query

```ts
import { createModel } from 'xstate/lib/model';

interface Data {
  id: string;
}

const model = createModel({}).withInput(
  {} as {
    data?: Data;
    error?: string;
  }
);

const machine = model.createMachine(
  {
    initial: 'pending',
    always: [
      {
        // If we have an error on the input, go to the hasErrored state
        cond: 'hasErrorOnInput',
        target: 'hasErrored'
      },
      {
        // If we have data on the input, go to the hasData state
        cond: 'hasDataOnInput',
        target: 'hasData'
      }
    ],
    states: {
      pending: {},
      hasData: {},
      hasErrored: {}
    }
  },
  {
    guards: {
      hasDataOnInput: (context, event, meta) => {
        return Boolean(meta.state.input.data);
      },
      hasErrorOnInput: (context, event, meta) => {
        return Boolean(meta.state.input.error);
      }
    }
  }
);
```

## Quick Reference

> #### Declare it in the options object

Author: Yes
```suggestion
#### Declare input in the options object
```

```ts
const machine = createMachine(
  {},
  {
    input: {
      foo: 'bar'
    }
  }
);
```

> #### Declare it using .withConfig

Author: ```suggestion
#### Redeclare input using .withConfig
```

```ts
const newMachine = machine.withConfig({
  input: {
    foo: 'bar'
  }
});
```

> #### Update it in a running service with `.input`

Author: ```suggestion
#### Update input in a running service with `.input`
```

```ts
const service = interpret(machine).start();

/**
 * It doesn't have to be the whole input -
 * it can be a partial
 */
service.input({
  foo: 'bar'
});
```

#### Listening for input changes

```ts
const machine = createMachine({
  on: {
    'xstate.input': {
      actions: (context, event, meta) => {
        console.log(meta.state.input);
      }
    }
  }
});
```

> #### Declare its type using `createModel`

Author: ```suggestion
#### Declare the input type using `createModel`
```
